File: backend-core/src/routes/notifications.js
const express = require('express');
const router = express.Router();
const { query } = require('../config/db');
const { requireAuth } = require('../middleware/auth');

router.use(requireAuth);

// GET /api/notifications
router.get('/', async (req, res, next) => {
  try {
    const notifsRes = await query(
      `SELECT id, type, title, body, link, is_read, metadata, created_at
       FROM notifications
       WHERE user_id = $1
       ORDER BY created_at DESC
       LIMIT 50`,
      [req.user.id]
    );

    const unreadRes = await query(
      'SELECT COUNT(*) AS count FROM notifications WHERE user_id = $1 AND is_read = false',
      [req.user.id]
    );

    res.json({
      success: true,
      data: {
        notifications: notifsRes.rows,
        unread_count: parseInt(unreadRes.rows[0].count, 10)
      }
    });
  } catch (err) {
    next(err);
  }
});

// PUT /api/notifications/:id/read
router.put('/:id/read', async (req, res, next) => {
  try {
    const { id } = req.params;
    await query(
      'UPDATE notifications SET is_read = true WHERE id = $1 AND user_id = $2',
      [id, req.user.id]
    );
    res.json({ success: true, message: 'Notification marked as read.' });
  } catch (err) {
    next(err);
  }
});

// PUT /api/notifications/read-all
router.put('/read-all', async (req, res, next) => {
  try {
    await query(
      'UPDATE notifications SET is_read = true WHERE user_id = $1 AND is_read = false',
      [req.user.id]
    );
    res.json({ success: true, message: 'All notifications marked as read.' });
  } catch (err) {
    next(err);
  }
});

module.exports = router;

File: backend-core-py/app/routers/notifications.py
import json
from fastapi import APIRouter, Depends
from app.auth.dependencies import require_auth
from app.db import get_db

router = APIRouter(tags=["notifications"])


@router.get("")
@router.get("/")
async def get_notifications(
    user: dict = Depends(require_auth),
    db=Depends(get_db),
) -> dict:
    rows = await db.fetch(
        """SELECT id, type, title, body, link, is_read, metadata, created_at
           FROM notifications
           WHERE user_id = $1
           ORDER BY created_at DESC
           LIMIT 50""",
        user["id"],
    )

    unread_count = await db.fetchval(
        "SELECT COUNT(*) FROM notifications WHERE user_id = $1 AND is_read = false",
        user["id"],
    )

    notifications = []
    for r in rows:
        d = dict(r)
        d["id"] = str(d["id"])
        if d.get("created_at"):
            d["created_at"] = d["created_at"].isoformat()
        if isinstance(d.get("metadata"), str):
            try:
                d["metadata"] = json.loads(d["metadata"])
            except Exception:
                pass
        notifications.append(d)

    return {
        "success": True,
        "data": {
            "notifications": notifications,
            "unread_count": int(unread_count or 0),
        },
    }


@router.put("/{notif_id}/read")
async def mark_notification_read(
    notif_id: str,
    user: dict = Depends(require_auth),
    db=Depends(get_db),
) -> dict:
    await db.execute(
        "UPDATE notifications SET is_read = true WHERE id = $1 AND user_id = $2",
        notif_id,
        user["id"],
    )
    return {"success": True, "message": "Notification marked as read."}


@router.put("/read-all")
async def mark_all_notifications_read(
    user: dict = Depends(require_auth),
    db=Depends(get_db),
) -> dict:
    await db.execute(
        "UPDATE notifications SET is_read = true WHERE user_id = $1 AND is_read = false",
        user["id"],
    )
    return {"success": True, "message": "All notifications marked as read."}
